File: provider/terraform_oics.rb
# Copyright 2017 Google Inc.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'provider/terraform'

module Provider
  # Code generator for runnable Terraform examples that can be run via an
  # Open in Cloud Shell link.
  class TerraformOiCS < Provider::Terraform
    # We don't want *any* static generation, so we override generate to only
    # generate objects.
    def generate(output_folder, types, _product_path, _dump_yaml)
      generate_objects(output_folder, types)
    end

    # Create a directory of examples per resource
    def generate_resource(data)
      examples = data.object.examples
                     .reject(&:skip_test)
                     .reject { |e| !e.test_env_vars.nil? && e.test_env_vars.any? }
                     .reject { |e| @version < @api.version_obj_or_closest(e.min_version) }

      examples.each do |example|
        target_folder = data.output_folder
        target_folder = File.join(target_folder, example.name)
        FileUtils.mkpath target_folder

        data.example = example

        data.generate(
          'templates/terraform/examples/base_configs/example_file.tf.erb',
          File.join(target_folder, 'main.tf'),
          self
        )

        data.generate(
          'templates/terraform/examples/base_configs/tutorial.md.erb',
          File.join(target_folder, 'tutorial.md'),
          self
        )

        data.generate(
          'templates/terraform/examples/base_configs/example_backing_file.tf.erb',
          File.join(target_folder, 'backing_file.tf'),
          self
        )

        data.generate(
          'templates/terraform/examples/static/motd',
          File.join(target_folder, 'motd'),
          self
        )
      end
    end

    # We don't want to generate anything but the resource.
    def generate_resource_tests(data) end
  end
end

File: mmv1/provider/terraform_oics.go
// Copyright 2024 Google Inc.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Code generator for a library converting terraform state to gcp objects.

package provider

import (
	"fmt"
	"log"
	"os"
	"path"
	"time"

	"github.com/GoogleCloudPlatform/magic-modules/mmv1/api"
	"github.com/GoogleCloudPlatform/magic-modules/mmv1/api/product"
)

type TerraformOiCS struct {
	TargetVersionName string

	Version product.Version

	Product *api.Product

	StartTime time.Time
}

func NewTerraformOiCS(product *api.Product, versionName string, startTime time.Time) TerraformOiCS {
	toics := TerraformOiCS{
		Product:           product,
		TargetVersionName: versionName,
		Version:           *product.VersionObjOrClosest(versionName),
		StartTime:         startTime,
	}

	toics.Product.SetPropertiesBasedOnVersion(&toics.Version)

	return toics
}

func (toics TerraformOiCS) Generate(outputFolder, productPath, resourceToGenerate string, generateCode, generateDocs bool) {
	toics.GenerateObjects(outputFolder, resourceToGenerate, generateCode, generateDocs)
}

func (toics TerraformOiCS) GenerateObjects(outputFolder, resourceToGenerate string, generateCode, generateDocs bool) {
	for _, object := range toics.Product.Objects {
		object.ExcludeIfNotInVersion(&toics.Version)

		if resourceToGenerate != "" && object.Name != resourceToGenerate {
			log.Printf("Excluding %s per user request", object.Name)
			continue
		}

		toics.GenerateObject(*object, outputFolder, toics.TargetVersionName, generateCode, generateDocs)
	}
}

func (toics TerraformOiCS) GenerateObject(object api.Resource, outputFolder, resourceToGenerate string, generateCode, generateDocs bool) {
	templateData := NewTemplateData(outputFolder, toics.TargetVersionName)

	if !object.IsExcluded() {
		log.Printf("Generating %s resource", object.Name)
		toics.GenerateResource(object, *templateData, outputFolder, generateCode, generateDocs)
	}
}

func (toics TerraformOiCS) GenerateResource(object api.Resource, templateData TemplateData, outputFolder string, generateCode, generateDocs bool) {
	if !generateDocs {
		return
	}

	for _, example := range object.TestExamples() {
		if len(example.TestEnvVars) > 0 {
			continue
		}

		example.SetOiCSHCLText()

		targetFolder := path.Join(outputFolder, example.Name)

		if err := os.MkdirAll(targetFolder, os.ModePerm); err != nil {
			log.Println(fmt.Errorf("error creating oics example directory %v: %v", targetFolder, err))
		}

		oicsExampleTemplatePath := "templates/terraform/examples/base_configs/oics_example_file.tf.tmpl"
		oicsExampleTemplates := []string{
			oicsExampleTemplatePath,
		}
		templateData.GenerateFile(path.Join(targetFolder, "main.tf"), oicsExampleTemplatePath, example, false, oicsExampleTemplates...)

		tutorialTemplatePath := "templates/terraform/examples/base_configs/tutorial.md.tmpl"
		tutorialTemplates := []string{
			tutorialTemplatePath,
		}
		templateData.GenerateFile(path.Join(targetFolder, "tutorial.md"), tutorialTemplatePath, example, false, tutorialTemplates...)

		backingTemplatePath := "templates/terraform/examples/base_configs/example_backing_file.tf.tmpl"
		backingTemplates := []string{
			backingTemplatePath,
		}
		templateData.GenerateFile(path.Join(targetFolder, "backing_file.tf"), backingTemplatePath, example, false, backingTemplates...)

		motdTemplatePath := "templates/terraform/examples/static/motd.tmpl"
		motdTemplates := []string{
			motdTemplatePath,
		}
		templateData.GenerateFile(path.Join(targetFolder, "motd"), motdTemplatePath, example, false, motdTemplates...)
	}
}

func (toics TerraformOiCS) CompileCommonFiles(outputFolder string, products []*api.Product, overridePath string) {

}

func (toics TerraformOiCS) CopyCommonFiles(outputFolder string, generateCode, generateDocs bool) {

}
